File: android/src/main/kotlin/com/orangecloud/im/flutter/OrangecloudImClientPlugin.kt
package com.orangecloud.im.flutter

import android.content.Context
import com.google.gson.Gson
import com.orangecloud.im.IMConnectionState
import com.orangecloud.im.OrangeCloudIMClient
import com.orangecloud.im.models.GiftInfo
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

/**
 * OrangeCloud IM Flutter 插件（Android 侧）
 *
 * 桥接 Dart 层与原生核心 SDK（com.orangecloud.im.OrangeCloudIMClient，AAR 二进制）：
 * - MethodChannel `com.orangecloud.im/methods`：接收 Dart 下发的命令
 * - EventChannel  `com.orangecloud.im/events`：把核心 SDK 的事件推送给 Dart
 *
 * 所有 IM 核心逻辑（连接/去重/补发/重连/心跳）均在 AAR 中，本类不含业务逻辑。
 */
class OrangecloudImClientPlugin : FlutterPlugin, MethodChannel.MethodCallHandler,
    EventChannel.StreamHandler {

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null
    private var appContext: Context? = null

    private val gson = Gson()
    private val client = OrangeCloudIMClient()
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
    private val collectJobs = mutableListOf<Job>()

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        appContext = binding.applicationContext
        methodChannel = MethodChannel(binding.binaryMessenger, "com.orangecloud.im/methods")
        methodChannel.setMethodCallHandler(this)
        eventChannel = EventChannel(binding.binaryMessenger, "com.orangecloud.im/events")
        eventChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        cancelCollectors()
        client.dispose()
        scope.cancel()
        appContext = null
    }

    // === EventChannel ===

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
        subscribeClientEvents()
    }

    override fun onCancel(arguments: Any?) {
        cancelCollectors()
        eventSink = null
    }

    private fun emit(event: String, data: Any?) {
        scope.launch {
            eventSink?.success(mapOf("event" to event, "data" to data))
        }
    }

    private fun subscribeClientEvents() {
        cancelCollectors()
        collectJobs += scope.launch {
            client.onConnectionStateChanged.collect { emit("connectionStateChanged", it.toWire()) }
        }
        collectJobs += scope.launch {
            client.onTextMessage.collect { emit("messageReceived", it.toJson()) }
        }
        collectJobs += scope.launch {
            client.onGiftMessage.collect { emit("messageReceived", it.toJson()) }
        }
        collectJobs += scope.launch {
            client.onSystemNotice.collect { emit("messageReceived", it.toJson()) }
        }
        collectJobs += scope.launch {
            client.onCustomMessage.collect { emit("messageReceived", it.toJson()) }
        }
        collectJobs += scope.launch {
            client.onBroadcastReceived.collect { emit("broadcastReceived", it.toJson()) }
        }
        collectJobs += scope.launch {
            client.onBatchMessageReceived.collect { list ->
                emit("batchMessageReceived", list.map { it.toJson() })
            }
        }
        collectJobs += scope.launch {
            client.onStateRestored.collect {
                emit("stateRestored", mapOf(
                    "restoredGroupIds" to it.restoredGroupIds,
                    "backfilledMessageCount" to it.backfilledMessageCount
                ))
            }
        }
        collectJobs += scope.launch {
            client.onReconnectAttempt.collect {
                emit("reconnectAttempt", mapOf(
                    "attemptNumber" to it.attemptNumber,
                    "nextDelayMs" to it.nextDelayMs
                ))
            }
        }
        collectJobs += scope.launch {
            client.onReconnectFailed.collect { emit("reconnectFailed", null) }
        }
        collectJobs += scope.launch {
            client.onGroupMemberList.collect { emit("groupMemberList", it) }
        }
    }

    private fun cancelCollectors() {
        collectJobs.forEach { it.cancel() }
        collectJobs.clear()
    }

    // === MethodChannel ===

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        try {
            when (call.method) {
                "login" -> {
                    client.maxReconnectAttempts = call.argument<Int>("maxReconnectAttempts") ?: -1
                    appContext?.let { client.registerNetworkCallback(it) }
                    client.login(
                        call.argument<String>("hubUrl") ?: "",
                        call.argument<String>("appId") ?: "",
                        call.argument<String>("userId") ?: "",
                        call.argument<String>("userSig") ?: ""
                    )
                    result.success(null)
                }
                "logout" -> { client.logout(); result.success(null) }
                "joinGroup" -> {
                    client.joinGroup(call.argument<String>("groupId") ?: "")
                    result.success(null)
                }
                "quitGroup" -> {
                    client.quitGroup(call.argument<String>("groupId") ?: "")
                    result.success(null)
                }
                "sendGroupMsg" -> {
                    client.sendGroupMsg(
                        call.argument<String>("groupId") ?: "",
                        call.argument<String>("messageJson") ?: ""
                    )
                    result.success(null)
                }
                "sendTextMessage" -> {
                    client.sendTextMessage(
                        call.argument<String>("groupId") ?: "",
                        call.argument<String>("content") ?: ""
                    )
                    result.success(null)
                }
                "sendGiftMessage" -> {
                    @Suppress("UNCHECKED_CAST")
                    val g = call.argument<Map<String, Any?>>("giftInfo") ?: emptyMap()
                    client.sendGiftMessage(
                        call.argument<String>("groupId") ?: "",
                        GiftInfo(
                            giftId = g["giftId"]?.toString() ?: "",
                            giftName = g["giftName"]?.toString() ?: "",
                            giftCount = (g["giftCount"] as? Number)?.toInt() ?: 1,
                            giftPrice = (g["giftPrice"] as? Number)?.toInt() ?: 0,
                            animationUrl = g["animationUrl"]?.toString() ?: ""
                        )
                    )
                    result.success(null)
                }
                "sendCustomMessage" -> {
                    @Suppress("UNCHECKED_CAST")
                    val payload = call.argument<Map<String, Any>>("payload") ?: emptyMap()
                    client.sendCustomMessage(
                        call.argument<String>("groupId") ?: "",
                        call.argument<String>("customType") ?: "",
                        payload
                    )
                    result.success(null)
                }
                "getGroupMemberList" -> {
                    client.getGroupMemberList(call.argument<String>("groupId") ?: "")
                    result.success(null)
                }
                "getLastSequenceNumber" -> {
                    result.success(
                        client.getLastSequenceNumber(call.argument<String>("groupId") ?: "")
                    )
                }
                "dispose" -> { client.dispose(); result.success(null) }
                else -> result.notImplemented()
            }
        } catch (e: Exception) {
            result.error("IM_ERROR", e.message, null)
        }
    }

    private fun com.orangecloud.im.models.IMMessage.toJson(): String = gson.toJson(this)

    private fun IMConnectionState.toWire(): String = when (this) {
        IMConnectionState.DISCONNECTED -> "disconnected"
        IMConnectionState.CONNECTING -> "connecting"
        IMConnectionState.CONNECTED -> "connected"
        IMConnectionState.RECONNECTING -> "reconnecting"
        IMConnectionState.RESTORING -> "restoring"
    }
}

File: ios/Classes/OrangecloudImClientPlugin.swift
import Flutter
import Foundation
import Combine
import OrangeCloudIMClient

/// OrangeCloud IM Flutter 插件（iOS 侧）
///
/// 桥接 Dart 层与原生核心 SDK（OCIMClient，XCFramework 二进制）：
/// - MethodChannel `com.orangecloud.im/methods`：接收 Dart 命令
/// - EventChannel  `com.orangecloud.im/events`：把核心 SDK 事件推给 Dart
///
/// IM 核心逻辑全部在 XCFramework 中，本类不含业务逻辑。
public class OrangecloudImClientPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {

    private let client = OCIMClient()
    private var eventSink: FlutterEventSink?
    private var bag = Set<AnyCancellable>()

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = OrangecloudImClientPlugin()
        let methodChannel = FlutterMethodChannel(
            name: "com.orangecloud.im/methods", binaryMessenger: registrar.messenger())
        let eventChannel = FlutterEventChannel(
            name: "com.orangecloud.im/events", binaryMessenger: registrar.messenger())
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        eventChannel.setStreamHandler(instance)
    }

    // MARK: - EventChannel

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        subscribeClientEvents()
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        bag.removeAll()
        eventSink = nil
        return nil
    }

    private func emit(_ event: String, _ data: Any?) {
        eventSink?(["event": event, "data": data as Any])
    }

    /// 把核心 SDK 的 Codable 消息编码为 JSON 字符串（与 data 下扁平字段契约一致）
    private static func encode<T: Encodable>(_ value: T) -> String? {
        guard let data = try? JSONEncoder().encode(value) else { return nil }
        return String(data: data, encoding: .utf8)
    }

    private func subscribeClientEvents() {
        bag.removeAll()

        client.onConnectionStateChanged.sink { [weak self] in
            self?.emit("connectionStateChanged", $0.rawValue)
        }.store(in: &bag)

        client.onTextMessage.sink { [weak self] in self?.emit("messageReceived", Self.encode($0)) }.store(in: &bag)
        client.onGiftMessage.sink { [weak self] in self?.emit("messageReceived", Self.encode($0)) }.store(in: &bag)
        client.onSystemNotice.sink { [weak self] in self?.emit("messageReceived", Self.encode($0)) }.store(in: &bag)
        client.onCustomMessage.sink { [weak self] in self?.emit("messageReceived", Self.encode($0)) }.store(in: &bag)

        client.onBroadcastReceived.sink { [weak self] in self?.emit("broadcastReceived", Self.encode($0)) }.store(in: &bag)
        client.onBatchMessageReceived.sink { [weak self] msgs in
            self?.emit("batchMessageReceived", msgs.compactMap { Self.encode($0) })
        }.store(in: &bag)
        client.onStateRestored.sink { [weak self] info in
            self?.emit("stateRestored", [
                "restoredGroupIds": info.groupIds,
                "backfilledMessageCount": info.backfilledCount
            ])
        }.store(in: &bag)
        client.onReconnectAttempt.sink { [weak self] info in
            self?.emit("reconnectAttempt", [
                "attemptNumber": info.attemptNumber,
                "nextDelayMs": info.delayMs
            ])
        }.store(in: &bag)
        client.onReconnectFailed.sink { [weak self] in self?.emit("reconnectFailed", nil) }.store(in: &bag)
        client.onGroupMemberList.sink { [weak self] in self?.emit("groupMemberList", $0) }.store(in: &bag)
    }

    // MARK: - MethodChannel

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [:]
        switch call.method {
        case "login":
            client.maxReconnectAttempts = args["maxReconnectAttempts"] as? Int ?? -1
            client.login(
                hubUrl: args["hubUrl"] as? String ?? "",
                appId: args["appId"] as? String ?? "",
                userId: args["userId"] as? String ?? "",
                userSig: args["userSig"] as? String ?? ""
            )
            result(nil)
        case "logout":
            client.logout(); result(nil)
        case "joinGroup":
            client.joinGroup(args["groupId"] as? String ?? ""); result(nil)
        case "quitGroup":
            client.quitGroup(args["groupId"] as? String ?? ""); result(nil)
        case "sendGroupMsg":
            client.sendGroupMsg(
                groupId: args["groupId"] as? String ?? "",
                messageJson: args["messageJson"] as? String ?? "")
            result(nil)
        case "sendTextMessage":
            client.sendTextMessage(
                groupId: args["groupId"] as? String ?? "",
                content: args["content"] as? String ?? "")
            result(nil)
        case "sendGiftMessage":
            let g = args["giftInfo"] as? [String: Any] ?? [:]
            client.sendGiftMessage(
                groupId: args["groupId"] as? String ?? "",
                giftInfo: GiftInfo(
                    giftId: g["giftId"] as? String ?? "",
                    giftName: g["giftName"] as? String ?? "",
                    giftCount: g["giftCount"] as? Int ?? 1,
                    giftPrice: g["giftPrice"] as? Int ?? 0,
                    animationUrl: g["animationUrl"] as? String ?? ""))
            result(nil)
        case "sendCustomMessage":
            client.sendCustomMessage(
                groupId: args["groupId"] as? String ?? "",
                customType: args["customType"] as? String ?? "",
                payload: args["payload"] as? [String: Any] ?? [:])
            result(nil)
        case "getGroupMemberList":
            client.getGroupMemberList(args["groupId"] as? String ?? ""); result(nil)
        case "getLastSequenceNumber":
            result(client.getLastSequenceNumber(groupId: args["groupId"] as? String ?? ""))
        case "dispose":
            client.logout(); result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}
